provisioner/terraform/tfparse/tfextract.go

```diff
@@ -88,16 +112,280 @@ func WorkspaceTags(ctx context.Context, logger slog.Logger, module *tfconfig.Mod
+			return nil, cleanup, xerrors.Errorf("convert zip to tar: %w", err)
+		}
+		rdr = bytes.NewReader(tarBytes)
+	default:
+		return nil, cleanup, xerrors.Errorf("unsupported mimetype: %s", mimetype)
+	}
+
+	if err := provisionersdk.Untar(tmpDir, rdr); err != nil {
+		return nil, cleanup, xerrors.Errorf("untar: %w", err)
+	}
+
+	module, diags := tfconfig.LoadModule(tmpDir)
+	if diags.HasErrors() {
+		return nil, cleanup, xerrors.Errorf("load module: %s", diags.Error())
+	}
+
+	return module, cleanup, nil
+}
+
+// loadVarsDefaults returns the default values for all variables passed to it.
+func loadVarsDefaults(ctx context.Context, logger slog.Logger, variables []*tfconfig.Variable) (map[string]string, error) {
+	// iterate through vars to get the default values for all
+	// variables.
+	m := make(map[string]string)
+	for _, v := range variables {
+		if v == nil {
+			continue
+		}
+		sv, err := interfaceToString(v.Default)
+		if err != nil {
+			return nil, xerrors.Errorf("can't convert variable default value to string: %v", err)
+		}
+		m[v.Name] = strings.Trim(sv, `"`)
+	}
+	logger.Debug(ctx, "found default values for variables", slog.F("defaults", m))
+	return m, nil
+}
+
+// loadParamsDefaults returns the default values of all coder_parameter data sources data sources provided.
+func loadParamsDefaults(ctx context.Context, logger slog.Logger, parser Parser, dataSources []*tfconfig.Resource) (map[string]string, error) {
+	defaultsM := make(map[string]string)
+	var skipped []string
+	for _, dataResource := range dataSources {
+		if dataResource == nil {
+			continue
+		}
+
+		if dataResource.Type != "coder_parameter" {
+			skipped = append(skipped, strings.Join([]string{"data", dataResource.Type, dataResource.Name}, "."))
+			continue
+		}
+
+		var file *hcl.File
+		var diags hcl.Diagnostics
+
+		if !strings.HasSuffix(dataResource.Pos.Filename, ".tf") {
+			continue
+		}
+
+		// We know in which HCL file is the data resource defined.
+		file, diags = parser.ParseHCLFile(dataResource.Pos.Filename)
+		if diags.HasErrors() {
+			return nil, xerrors.Errorf("can't parse the resource file %q: %s", dataResource.Pos.Filename, diags.Error())
+		}
+
+		// Parse root to find "coder_parameter".
+		content, _, diags := file.Body.PartialContent(rootTemplateSchema)
+		if diags.HasErrors() {
+			return nil, xerrors.Errorf("can't parse the resource file: %s", diags.Error())
+		}
+
+		// Iterate over blocks to locate the exact "coder_parameter" data resource.
+		for _, block := range content.Blocks {
+			if !slices.Equal(block.Labels, []string{"coder_parameter", dataResource.Name}) {
+				continue
+			}
+
+			// Parse "coder_parameter" to find the default value.
+			resContent, _, diags := block.Body.PartialContent(coderParameterSchema)
+			if diags.HasErrors() {
+				return nil, xerrors.Errorf(`can't parse the coder_parameter: %s`, diags.Error())
+			}
+
+			if _, ok := resContent.Attributes["default"]; !ok {
+				defaultsM[dataResource.Name] = ""
```

Comment:
Should we specify if the default is `""` or actually not present?

This would make a tag `""` if no default is provided. Should we error instead? Is that a breaking change if we error?

Reply:
That's a good point. The whole reason for doing this is to make sure that we have _some_ default defined for the `workspace_tags` data source. So it probably should be an error.

Reply:
Upon closer inspection, this is _only_ for parsing `coder_parameters`. It is legit to have an empty default value for a `coder_parameter`. We do not want to block this behaviour.
However, when using that `coder_parameter` as a transitive source for a `coder_workspace_tags` value, we want to ensure that that tag is set. This is checked in `WorkspaceTagDefaults` and has its own corresponding unit test.